File: src/main/java/me/urielsalis/mojiraFeedIRC/IRCServer.java
package me.urielsalis.mojiraFeedIRC;

import nedhyett.Amelia.Amelia;
import nedhyett.Amelia.CommandRegistry;
import nedhyett.Amelia.core.config.ConfigReader;
import nedhyett.Amelia.core.connection.ConnectionThread;
import nedhyett.Amelia.core.users.User;
import nedhyett.Amelia.managers.UserManager;
import nedhyett.Amelia.ping.PingPongThread;
import nedhyett.Amelia.services.NickServ;
import nedhyett.crimson.logging.CrimsonLog;

import java.io.*;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

/**
 * mojiraFeedIRC
 * <p>
 * Created by urielsalis on 09/08/17.
 */
public class IRCServer extends Amelia {
    public static HashMap<String, UserFeed> userList = new HashMap<String, UserFeed>();

    public IRCServer(String[] args) {
        CrimsonLog.initialise("IRC");
        this.cmdargs = args;

        CrimsonLog.info("Loading configuration...");
        ConfigReader cr = new ConfigReader("config.txt");
        cr.read();
        config = cr.parse();
        CrimsonLog.info("Configuration loaded!");

        CrimsonLog.info("Loading MOTD...");
        File motdf = new File("motd.conf");
        if (motdf.exists()) {
            try {
                BufferedReader br = new BufferedReader(new InputStreamReader(new FileInputStream(motdf)));
                String line = br.readLine();
                while (line != null) {
                    MOTD.add(line);
                    line = br.readLine();
                }
                CrimsonLog.info("MOTD loaded!");
            } catch (IOException e) {
                MOTD.clear();
                CrimsonLog.warning("IOException while loading MOTD!");
            }
        } else {
            CrimsonLog.warning("motd.conf not found!");
        }

        CrimsonLog.info("Starting connection thread...");
        ConnectionThread ct = new ConnectionThread();
        ct.start();
        CrimsonLog.info("Connection thread started!");

        CrimsonLog.info("Starting PingPong thread...");
        PingPongThread ppt = new PingPongThread();
        ppt.start();
        CrimsonLog.info("PingPong thread started!");

        UserManager.addFakeUser("FeedServer", new FeedServer("FeedServer", "FeedServer", "Urielsalis", "FeedServer@Urielsalis"));

        Amelia.startupTime = new Date();

    }

    public void newFeed(Feed feedObj) {
        loop: for(Map.Entry<String, UserFeed> entry: userList.entrySet()) {
            UserFeed feed = entry.getValue();
            if(feed.user!=null && feed.user.respondedToLastPing) {
                if(feedObj.author.toLowerCase().contains(feed.username.toLowerCase())) continue;
                for(String str: feed.ignoreList) {
                    if(feedObj.title.contains(str)) {
                        continue loop;
                    }
                }
                feed.user.sendRaw(feedObj.author, "PRIVMSG #feed :\u000307 "+feedObj.link+"\u000f - "+feedObj.title);
            }
        }
    }
}

File: src/main/java/me/urielsalis/mojiraFeedIRC/IRCServer.kt
package me.urielsalis.mojiraFeedIRC

import nedhyett.Amelia.Amelia
import nedhyett.Amelia.core.config.ConfigReader
import nedhyett.Amelia.core.connection.ConnectionThread
import nedhyett.Amelia.managers.ChannelManager
import nedhyett.Amelia.managers.UserManager
import nedhyett.Amelia.ping.PingPongThread
import nedhyett.crimson.logging.CrimsonLog
import nedhyett.crimson.logging.CrimsonLog.info
import nedhyett.crimson.logging.CrimsonLog.warning
import org.apache.commons.collections4.CollectionUtils
import java.io.*
import java.util.*

object IRCServer : Amelia() {
    var args: Array<String> = emptyArray()
    var userList: MutableMap<String, UserFeed> = mutableMapOf()

    fun start() {
        CrimsonLog.initialise("IRC")
        cmdargs = args

        info("Loading configuration...")
        val cr = ConfigReader("config.txt")
        cr.read()
        Amelia.config = cr.parse()
        info("Configuration loaded!")

        info("Loading MOTD...")
        val motdf = File("motd.conf")
        if (motdf.exists()) {
            try {
                val br = BufferedReader(InputStreamReader(FileInputStream(motdf)))
                var line: String? = br.readLine()
                while (line != null) {
                    Amelia.MOTD.add(line)
                    line = br.readLine()
                }
                info("MOTD loaded!")
            } catch (e: IOException) {
                Amelia.MOTD.clear()
                warning("IOException while loading MOTD!")
            }

        } else {
            warning("motd.conf not found!")
        }

        info("Starting connection thread...")
        val ct = ConnectionThread()
        ct.start()
        info("Connection thread started!")

        info("Starting PingPong thread...")
        val ppt = PingPongThread()
        ppt.start()
        info("PingPong thread started!")

        UserManager.addFakeUser("FeedServer", FeedServer("FeedServer", "FeedServer", "Urielsalis", "FeedServer@Urielsalis"))

        info("Setting OP in #feed")
        ChannelManager.createChannel("#feed")
        val feedChannel = ChannelManager.getChannel("#feed")
        val feedserver = UserManager.getUser("FeedServer")
        feedChannel.join(feedserver)
        feedChannel.op("FeedServer")

        info("Configuration done!")
        Amelia.startupTime = Date()
    }

    fun newFeed(feedObj: Feed) {
        val nulls = userList
                .mapValues {  UserManager.getUser(it.key) }
                .filterValues { it == null }
        nulls.forEach { userList.remove(it.key) }
        val inactive = userList
                .mapValues {  UserManager.getUser(it.key) }
                .filterValues { !it.respondedToLastPing }
        inactive.forEach { userList.remove(it.key)}
        userList
                .mapKeys {  UserManager.getUser(it.key) }
                .filterKeys { it != null }
                .filterKeys { it.respondedToLastPing }
                .filterKeys { !feedObj.author.contains(it.username, true) }
                .filterValues { feed -> !feed.ignoreList.any { feedObj.title.matches(it.toRegex()) } }
                .forEach { user, _ -> user.sendRaw(feedObj.author, "PRIVMSG #feed :\u000307${feedObj.link} \u000f- ${feedObj.title}") }

        refreshPerms()
    }

    private fun refreshPerms() {
        val feedChannel = ChannelManager.getChannel("#feed")
        userList
                .mapKeys {  UserManager.getUser(it.key) }
                .filterKeys { it == null || !it.respondedToLastPing }.forEach { username, _ -> feedChannel.devoice(username) }
    }
}
